core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -544,7 +587,105 @@ public GroupByIdBlock getResult()
             checkState(lastPosition == block.getPositionCount(), "process has not yet finished");
             checkState(!finished, "result has produced");
             finished = true;
-            return new GroupByIdBlock(nextGroupId, blockBuilder.build());
+            return new GroupByIdBlock(nextGroupId, new LongArrayBlock(groupIds.length, Optional.empty(), groupIds));
+        }
+    }
+
+    private void batchedPutIfAbsent(Block block, int blockOffset, int batchSize, long[] batchGroupIds, int batchGroupIdOffset)
```

Comment:
We don't care about small pages too much, but I wonder what is the impact of small blocks on performance.

Reply:
While doing actual queries from tpch there were some pages as small as 200 rows.
In general case there is no significant performance degradation. With dictionaries, however, this may change